### ISSUE TITLE
TWRP ignores DT fstab making it useless on non-standard installs

### ISSUE THREAD

**Issue by EatPrilosec** [twrp-raspberry/android_local_manifest#2]
the DT fstab overlays for sdcard/usb/nvme use /dev/block/<real device> while the fstab for twrp uses /dev/blcok/by-name/<partname>

i noticed in UART:

`init: First stage mount skipped (recovery mode)`

is there no way to enable [early partition mounting via DT ](https://source.android.com/docs/core/architecture/kernel/mounting-partitions-early#early-mount-dt-vboot1) for twrp? 

the issue is, the raspberry pi supports native multiboot, I have a setup where I've installed all three versions of your android builds, however, to make that work, I renamed the partitions from boot/system/vendor/userdata to aosp-boot/linos-sys/linatv-ven, respectively, and then i generate a DTBO for those mmcbllk0 partitions

but twrp seems to ignore the fstab overlay I've made despite twrp having the device tree loaded properly
 
![image](https://github.com/user-attachments/assets/e0a5f886-1c99-4b48-bc20-062a53cfa623)

if not possible to enable [early partition mounting via DT ](https://source.android.com/docs/core/architecture/kernel/mounting-partitions-early#early-mount-dt-vboot1), is there an init script that can be used to `sed -i` the fstab with the proper device paths from the device tree (by getting it from /proc/firmware/device-tree/android/fstab/<android partname>/dev) ?

maybe something like:

```
for partName in boot system vendor userdata
do
    partDev=$(cat /proc/firmware/device-tree/android/fstab/$partName/dev)
    [ ! -z "$partDev" ] && sed -i -e "s|/dev/block/by-name/$partName|$partDev|g" /system/etc/fstab
done
```

maybe here?: https://github.com/twrp-raspberry/android_bootable_recovery/blob/049c954f6bfb13c3917d1022c9ba67f640163065/etc/init.rc#L24

**Comment by EatPrilosec** [comment]
also, these additions to the OTA backup script (rpi-settings.sh) should enable OTA on non-standard boot configurations defined by user-defined DT fstabs


```
backup()
{
  #..... other commands

  if [ -f /boot/config_user.txt ]; then
    echo "Backup config_user.txt" >> /tmp/rpi-settings.log
    cp /boot/config_user.txt /tmp/config_user.txt

    # backup user-made dtbo defined in config_user.txt
    for overlay in $(grep -e "^dtoverlay=.*" /boot/config_user.txt | sed "s|dtoverlay=||" | grep -v "=")
    do
      skip=0
      for newOverlay in $(ls /tmp/overlays/*.dtbo | sed "s|/tmp/overlays/||" | sed "s|\.dtbo||")
      do
        [ "$overlay" = "$newOverlay" ] && skip=1
      done
      [ $skip = 1 ] && continue #not user defined so skip

      [ ! -d /tmp/overlay-user ] && mkdir /tmp/overlays-user

      [ -f "$overlaypath" ] && cp "/boot/overlays/${overlay}.dtbo" "/tmp/overlays-user/${overlay}.dtbo"
    done
  fi

  #..... other commands
}

restore()
{
  #..... other commands

  if [ -f /tmp/config_user.txt ]; then
    echo "Restore config_user.txt" >> /tmp/rpi-settings.log
    cp /tmp/config_user.txt /boot/config_user.txt

    # restore user-made dtbo defined in config_user.txt
    cp /tmp/overlays-user/* /boot/overlays/
  fi

  #..... other commands
}
```

**Comment by KonstaT** [comment]
> the DT fstab overlays for sdcard/usb/nvme use /dev/block/ while the fstab for twrp uses /dev/blcok/by-name/

That's what https://github.com/twrp-raspberry/android_system_core/commit/2df43473ea252735ebbb6f4c2a366c910ce61522 is for.

> the issue is, the raspberry pi supports native multiboot, ...

Raspberry Pi supports booting from various different storage devices (sdcard, USB, NVME). No, it does not natively support several OS'es on a single storage device.

I have no plans supporting any multi-boot solutions. I do need to support booting from various different storage devices (sdcard, USB, NVME).

While you could achieve something like this for your own personal builds with custom DT fstabs and modifying https://github.com/twrp-raspberry/android_system_core/commit/2df43473ea252735ebbb6f4c2a366c910ce61522, it's not something that's in scope of this project.

**Comment by EatPrilosec** [comment]
> > the DT fstab overlays for sdcard/usb/nvme use /dev/block/ while the fstab for twrp uses /dev/blcok/by-name/
> 
> That's what [twrp-raspberry/android_system_core@2df4347](https://github.com/twrp-raspberry/android_system_core/commit/2df43473ea252735ebbb6f4c2a366c910ce61522) is for.

noted, thank you.


> > the issue is, the raspberry pi supports native multiboot, ...
> 
> Raspberry Pi supports booting from various different storage devices (sdcard, USB, NVME). No, it does not natively support several OS'es on a single storage device.

**this is absolutely incorrect.** the official, raspberry pi foundation NOOBS installer (https://github.com/raspberrypi/noobs) and the modification of it called PINN both use this function of the RPI bootloader to boot a different partition on next boot after installing OSs. this is done by the cmdline arg `sdhci.debug_quirks2=4` and in linux  you can type `reboot <partition>` to reboot to that partiton on the current boot device (regardless of whether that device is sdcard/usb/nvme), and the eeprom then treats that as the boot partiton and reads config.txt and cmdline.txt, kernel etc.

its talked about in great detail here 
https://github.com/raspberrypi/noobs/issues/635

> I have no plans supporting any multi-boot solutions. I do need to support booting from various different storage devices (sdcard, USB, NVME).

i wholeheartedly disagree with this decision. the goal of the TWRP for your android versions should be to follow the rpi boot standards as closely as the android installation its paired to at the very least, DT fstab should override the four main entries in the fstab

> While you could achieve something like this for your own personal builds with custom DT fstabs and modifying [twrp-raspberry/android_system_core@2df4347](https://github.com/twrp-raspberry/android_system_core/commit/2df43473ea252735ebbb6f4c2a366c910ce61522), it's not something that's in scope of this project.

**Comment by EatPrilosec** [comment]
https://github.com/twrp-raspberry/android_system_core/commit/2df43473ea252735ebbb6f4c2a366c910ce61522

not to mention it would clean up this (self proclaimed) hack exponentially. instead of checking three sets of devices and  4 sets of partitions, you just make a symlink if `/proc/.../fstab/<mount>` exists for 4 mounts

**Comment by KonstaT** [comment]
> **this is absolutely incorrect.**

Comment by a Pi engineer in 2021, "We no longer recommend NOOBS, and in fact we have done no development on it for years."
https://forums.raspberrypi.com/viewtopic.php?t=318540#p1906632

PINN is a third party solution.

> would you reconsider if i did the work and made a pull request?

Like mentioned, there's no plans for supporting any multi-boot solutions.

> not to mention it would clean up this (self proclaimed) hack exponentially.

By introducing even worse hacks? I need to take into consideration what works for both Android and recovery mode.

If you're only interested in TWRP, you can just ignore the DT fstabs, the linked patch to create symlinks to `/dev/block/by-name/*`, and just directly modify the recovery fstab (https://github.com/twrp-raspberry/android_device_brcm_rpi/blob/twrp-11/recovery.fstab).

**Comment by EatPrilosec** [comment]
> > **this is absolutely incorrect.**
> 
> Comment by a Pi engineer in 2021, "We no longer recommend NOOBS, and in fact we have done no development on it for years." https://forums.raspberrypi.com/viewtopic.php?t=318540#p1906632
> 

The `sdhci.debug_quirks2=4` is baked into the rpi firmware, and the kernel. I'm pretty sure it works from ADB, let alone official raspi os and every other distro with the raspi kernel.

You can do `reboot <partnum>` from:
- RaspiOS 64
- Raspi OS legacy
- Ubuntu
- Kali
- Debian
- Arabian
- arch
- lakka
- libreelec
- ombc
- batocera
- recalbox

It's native, you were incorrect, and the fact that a 3rd party software ALSO uses that feature of the rpi bootloader doesn't make it non-native. try and accept this fact please.


> PINN is a third party solution.
> 

Immaterial, see above.

> > not to mention it would clean up this (self proclaimed) hack exponentially.
> 
> By introducing even worse hacks? I need to take into consideration what works for both Android and recovery mode.

Your hack is LITERALLY Worst case scenario 

It basically guesses, when you can get it from the 

DEVICE TREE.........

#like android does...

Which brings me to, how are you taking into consideration a method that works for both? Explain please.

> If you're only interested in TWRP, you can just ignore the DT fstabs, the linked patch to create symlinks to `/dev/block/by-name/*`, and just directly modify the recovery fstab (https://github.com/twrp-raspberry/android_device_brcm_rpi/blob/twrp-11/recovery.fstab).

This doesn't survive OTA, as not only you completely ignore custom overlays during the OTA process, you completely rewrite TWRP every OTA. 

If you could neuter the latter, and make the recovery ramdisk survive OTA, at the very least, then making a separate TWRP would be an acceptable solution..

I'm not trying to be rude, but you're being very obtuse.

**Comment by KonstaT** [comment]
> The sdhci.debug_quirks2=4 is baked into the rpi firmware, and the kernel. I'm pretty sure it works from ADB, let alone official raspi os and every other distro with the raspi kernel.

You're confusing Android with your regular Linux distro. reboot command in Android supports no such thing.
https://android.googlesource.com/platform/system/core/+/refs/tags/android-15.0.0_r1/reboot/reboot.c

> Your hack is LITERALLY Worst case scenario
> It basically guesses, when you can get it from the
> DEVICE TREE.........
> ....like android does...

Yeah, since you don't even understand how it currently works. Guess what, it gets it from the device tree! There's no guessing involved.

User select one of the device tree overlays to load in /boot/config.txt (i.e. dtoverlay=android-sdcard, dtoverlay=android-usb, or dtoverlay=android-nvme - https://github.com/raspberry-vanilla/android_kernel_brcm_rpi/commit/e8ebcb324bf2a65edb46774b1e39d56e17404916), https://github.com/twrp-raspberry/android_system_core/commit/2df43473ea252735ebbb6f4c2a366c910ce61522 symlinks the device partitions to /dev/block/by-name/*, Android/recovery mounts partition defined in fstab using dev/block/by-name/* paths. This is done for it to be possible to support booting from sdcard/USB/NVME devices.

Android has two stage init and there's actually two fstabs. One for first_stage mount partitions (system partitions) and another one for late mount partitions (userdata).

> This doesn't survive OTA, as not only you completely ignore custom overlays during the OTA process, you completely rewrite TWRP every OTA.

Yes, this does survive OTA just fine. And no, there's no reason to take any custom overlays into consideration during the OTA process as in practice no one has even added any custom overlays.

> I'm not trying to be rude, but you're being very obtuse.

Ok, since I'm clearly highly regarded, please ELI5 what was your non-hacky solution to this exactly.

Create unlimited amount of custom DT fstab overlays to cover every possible scenario (sdcard/USB/NVME boot devices and all partition numbers where the Android partition could end up in a multi-boot system) and hijack the init process with a script that will adjust fstab on thy fly?

**Comment by EatPrilosec** [comment]
> ELI5 what was your non-hacky solution to this exactly.

lets do this. nerd fight. huricane helen knocked out my power, sorry for late reply


> You're confusing Android with your regular Linux distro. Pretty sure reboot command in Android supports no such thing.

### *AHEM* 
https://streamable.com/36ydq5
lets start with a video of you being completely wrong. sorry. you are basically begging for this: it exires in two days
im not lying to you. its in the sdhci driver, apparantly. whether or not android is linux, its using that driver right? 

in this video you will see me
- boot YOUR aosp15 android distro and boot logo YES, DISTRO (because android is linux)
- in the bottom right, youll see the output of the pi in hdmi, the rest is a UART console to the various android consoles (because android is linux)
- then i will type blkid to see the partitions available on the system (because android is linux)
- then i type reboot 10, which will trigger a reboot and a console mesege that says something like `rebooting with command "10"`
-then, the system will reboot and you will see the bootlogo of your lineageOS DISTRO YES DISTRO. (because android is linux)

### you can try this yourself. it works. im not lying


>Create unlimited amount of custom DT fstab overlays to cover every possible scenario (sdcard/USB/NVME boot devices and all partition numbers where the Android partitions could end up in a multi-boot system) and hijack the init process with a script that will adjust fstab on the fly?

.....no, i make the dtbo at install time... the dtbo is not the issue, its that the ota doesnt make any effort to keep user made dtbo

 i need TWRP to READ THE DTBO, loaded into the device tree at boot, that contains fstab entries, added to the device tree, the actual entries in the dt. 

here is an example. i take this DTS here: https://github.com/raspberry-vanilla/android_kernel_brcm_rpi/blob/android-15.0/arch/arm/boot/dts/overlays/android-sdcard-overlay.dts. i have a script that changes the `dev=/dev/block/XXXXXXX`  from strings with the ones I specify, and adds a block for userdata: which turns into this:

```
/dts-v1/;
/plugin/;
/ {
    compatible = "brcm,bcm2835";
    fragment@0 {
        target-path = "/";
        __overlay__ {
            firmware {
                android {
                    compatible = "android,firmware";
                    bootdevice = "sdcard";
                    fstab {
                        compatible = "android,fstab";
                        boot {
                            compatible = "android,boot";
                            dev = "/dev/block/mmcblk0p10";
                            type = "vfat";
                            mnt_flags = "ro";
                            fsmgr_flags = "wait,first_stage_mount";
                        };
                        system {
                            compatible = "android,system";
                            dev = "/dev/block/mmcblk0p11";
                            type = "ext4";
                            mnt_flags = "ro,barrier=1";
                            fsmgr_flags = "wait,first_stage_mount";
                        };
                        vendor {
                            compatible = "android,vendor";
                            dev = "/dev/block/mmcblk0p12";
                            type = "ext4";
                            mnt_flags = "ro,barrier=1";
                            fsmgr_flags = "wait,first_stage_mount";
                        };
						data {
                            compatible = "android,data";
                            dev = "/dev/block/mmcblk0p13";
                            type = "ext4";
                            mnt_flags = "noatime,nosuid,nodev,nomblk_io_submit,errors=panic";
                            fsmgr_flags = "wait,check,formattable,quota";
                        };
                    };
                };
            };
        };
    };
};

```
i compile it with the device tree compiler as specified here: https://source.android.com/docs/core/architecture/kernel/mounting-partitions-early#early-mount-dt-vboot1

with a command that looks like this:
`dtc -@ -I dts -O dtb android-$bootdev-template-${1}${2}${3}${4}.dts -o android-$bootdev-${1}${2}${3}${4}.dtbo >/dev/null 2>&1`

i then place that custom overlay in the overlays folder of the 10th partition of the sdcard, named android-multiboot.dtbo

and add 

`dtoverlay=android-multiboot`

and then with no other modifications whatsoever, im able to install all three of your android distros to one install media following this process.


BUT, that again, is not the issue
were you even reading? i need your android to:

### one:
correctly handle user specified DTBO overlays in the OTA files, like i said in this comment: https://github.com/twrp-raspberry/android_local_manifest/issues/2#issuecomment-2378536439
not generate any. the USER is providing the fstab in the DTBO, i just need twrp to ADHERE TO THE DT OVERLAY I PROVIDED

i checked, TWRP's kernel loads this information into `/proc/firmware/device-tree/android/fstab/`
and you can get these custom dtbo DEV entries  from `/proc/firmware/device-tree/android/fstab/(partition)/dev` you can just `cat` them and in this case `cat /proc/firmware/device-tree/android/fstab/boot/dev` will return `/dev/block/mmcblk0p10`

have you heard of this command called `sed`? or??? 

so the hacky solution i propose woul look something like this: 
somewhere during the boot process, edit the fstab for TWRP, to use entires provided in the DT fstab by the user
```
# adhere to raspi DT-defined fstab, but use expected fstab values when no fstab in DT
    p1=$(cat /proc/firmware/device-tree/android/fstab/boot/dev)
    p2=$(cat /proc/firmware/device-tree/android/fstab/system/dev)
    p3=$(cat /proc/firmware/device-tree/android/fstab/vendor/dev)
    p4=$(cat /proc/firmware/device-tree/android/fstab/userdata/dev)

    [ -z "$p1" ] && p1=/dev/block/by-name/boot
    [ -z "$p2" ] && p2=/dev/block/by-name/system
    [ -z "$p3" ] && p3=/dev/block/by-name/vendor
    [ -z "$p4" ] && p4=/dev/block/by-name/userdata

    awk -v p1=$p1 -v p2=$p2 -v p3=$p3 -v p4=$p4 '
        BEGIN {FS=" "; OFS="\t";}
        /\/boot/   { $1=p1 }
        /\/system/ { $1=p2 }
        /\/vendor/ { $1=p3 }
        /\/data/   { $1=p4 }
        {print}
        ' /system/etc/recovery.fstab > /system/etc/recovery.fstab.new

    mv /system/etc/recovery.fstab.new /system/etc/recovery.fstab

```


### and two

correctly load that information in recovery mode, like the main android OS does as ove explained in this comment very thoroughly.



ok. deep breath.



### CURRENTLY, twrp MISHANDLES the raspberry pi boot process, and you yourself provided the evicence to that here: https://github.com/twrp-raspberry/android_system_core/commit/2df43473ea252735ebbb6f4c2a366c910ce61522


```
if ((uevent.device_name.compare("mmcblk0p1") == 0 && bootdevice.compare(0, 6, "sdcard") == 0)
            || (uevent.device_name.compare("sda1") == 0 && bootdevice.compare(0, 3, "usb") == 0)
            || (uevent.device_name.compare("nvme0n1p1") == 0 && bootdevice.compare(0, 4, "nvme") == 0)) {
        links.emplace_back("/dev/block/by-name/boot");
```


### this is using UEVENT 

### not dt 

### and ASSUMES THE FIRST PARTITION............ no matter what, 

hardcodded in c and compiled into the twrp binary, right?, 

rather than being user definable by a dtoverlay, like in android (See video)


>Android has two stage init and there's actually two fstabs. One for first_stage mount partitions (system partitions) and another one for late mount partitions (userdata).


CORRECT. that is why I'm in TWRP issues suggesting that TWRP be hacked, slightly better, to account for manually defined partitions in DT like I'm doing to multiboot android.


sorry for the attitude, i honestly have no clue why youre being so hostile. i dont think youre "highly reguarded" if that means what i think it means, and i am not trying to insult you, but rather, get twrp to handle the rpi boot process with more fidelity than it currently does, so i can make a multiboot android entertainment device for my lil bro.  you have the monopoly on rpi android builds. thats why i came here for your help. 

otas, and twrp, 

for the pi's bootloader native multiboot feature

thats all i'm going for here, and the changes are fairly small, and im willing to help with them!

**Comment by KonstaT** [comment]
All I've tried to do is to explain how this currently works, what it is intended to support, and what I need to take into consideration while doing this.

> Create unlimited amount of custom DT fstab overlays to cover every possible scenario (sdcard/USB/NVME boot devices and all partition numbers where the Android partitions could end up in a multi-boot system) and hijack the init process with a script that will adjust fstab on the fly?

We still didn't make much progress on this.

So, ok, something else somewhere else creates a DT fstab overlay that can contain any boot device and partition numbers.

Since fstab resides inside the ramdisk, it's not something you can just edit on thy fly. Only if there was some other way to hijack the init process. Like using uevent that goes through all the block devices to create some symlinks that can be used in the fstab that can't be edited directly?

If your magical `awk` skills translate into C++, you'd be at least one step closer.

**Comment by EatPrilosec** [comment]
> All I've tried to do is to explain how this currently works, what it is intended to support, and what I need to take into consideration while doing this.
> 
> > Create unlimited amount of custom DT fstab overlays to cover every possible scenario (sdcard/USB/NVME boot devices and all partition numbers where the Android partitions could end up in a multi-boot system) and hijack the init process with a script that will adjust fstab on the fly?
> 

> We still didn't make much progress on this.
no argument here.. 


> 
> So, ok, something else somewhere else creates a DT fstab overlay that can contain any boot device and partition numbers.

i feel like a broken record, but yes, 

### AS DEFINED IN THE ANDROID DOCUMENTATION
android can get a full fstab from a DT overlay, thats [LITERALLY how the full android boots ](https://github.com/raspberry-vanilla/android_kernel_brcm_rpi/blob/android-15.0/arch/arm/boot/dts/overlays/android-sdcard-overlay.dts)

the kernel loads the dtbo into `/proc/firmware/device-tree`

/proc/firmware/device-tree/android/fstab/boot/dev
/proc/firmware/device-tree/android/fstab/system/dev
/proc/firmware/device-tree/android/fstab/vendor/dev
/proc/firmware/device-tree/android/fstab/userdata/dev

and whatever elsle you decide to add to the dts before you (as in the user) compiles it

its the same place you get [your boot device variable](https://github.com/twrp-raspberry/android_system_core/commit/2df43473ea252735ebbb6f4c2a366c910ce61522#diff-a4fff1f4332968ca956a94fb33ee6af11ec716f4963e9b2c8f18d370802b572bR317) to test for the rpi hack you currently have in place.



> Since fstab resides inside the ramdisk, it's not something you can just edit on thy fly. Only if there was some other way to hijack the init process. Like using uevent that goes through all the block devices to create some symlinks that can be used in the fstab that can't be edited directly?
> 
> If your magical `awk` skills translate into C++, you'd be at least one step closer.

1) it being on the ramdisk does mean it can be edited on the fly, actually, it has write permissions, but youre correct, the PROPER way is to do it in th twrp binary.

2) ive written a draft and an trying to figure out how to compile twrp myself to test it. 

3) you: `All I've tried to do is to explain how this currently works`
also you `magical 'awk' skills`

i again do not understand where the attitude is coming from, do you think your code was magical? no. 

but anyway, yes, i dont have a lot of c++ experience, i have a lot of posix/shell experience which is why i suggested that, and also why, for the third time i think 


### I CAME TO YOU FOR HELP

holy crap. 

the current plan is to check for the existence of 
/proc/firmware/device-tree/android/fstab/boot/dev
/proc/firmware/device-tree/android/fstab/system/dev
/proc/firmware/device-tree/android/fstab/vendor/dev
/proc/firmware/device-tree/android/fstab/userdata/dev

if they exist, read the contents [like you have with bootdevice ](https://github.com/twrp-raspberry/android_system_core/commit/2df43473ea252735ebbb6f4c2a366c910ce61522#diff-a4fff1f4332968ca956a94fb33ee6af11ec716f4963e9b2c8f18d370802b572bR317) 

and then use those variables to link them

what im unsure about is how the function `links.emplace_back("/dev/block/by-name/boot");`  is getting what device to use [from the line above it](https://github.com/twrp-raspberry/android_system_core/commit/2df43473ea252735ebbb6f4c2a366c910ce61522#diff-a4fff1f4332968ca956a94fb33ee6af11ec716f4963e9b2c8f18d370802b572bR325). any help would certainly be appreciated

getting android to boot on the pi at all is a hack. i just want to help make twrp to use the same hack instead of a different one.

**Comment by KonstaT** [comment]
You're confusing Android and Android recovery which you can consider a separate mini-OS of its own.

As mentioned earlier, Android (not Android recovery) has two stage init. https://source.android.com/docs/core/architecture/kernel/mounting-partitions-early discusses mounting first_stage partitions. It also clearly states "Entries provided through device tree overlays must not repeat in the `fstab` file fragments. For example, when specifying an entry to mount` /vendor` in the devicetree, the `fstab` file must not repeat that entry.".

In case you were wondering why there's something missing from https://github.com/raspberry-vanilla/android_kernel_brcm_rpi/commit/e8ebcb324bf2a65edb46774b1e39d56e17404916 and Android fstab (again not Android recovery fstab) looks something like this:
```
/dev/block/by-name/userdata                /data        ext4    noatime,nosuid,nodev,nomblk_io_submit,errors=panic  wait,check,formattable,quota
```

If you were to add the userdata partition to the DT fstab with the correct mnt and fs_mgr flags, Android doesn't even boot. It needs to be tagged with recoveryonly so that the first stage init doesn't even attempt to mount it which is again a HACK.

https://github.com/twrp-raspberry/android_system_core/commit/2df43473ea252735ebbb6f4c2a366c910ce61522 is absolutely the correct place to create symlinks to /dev/block/by-name/*. That's where Android (and Android recovery) does it in any case. It's HACK because we're not getting the boot device from Android bootloader (since we don't have on Raspberry Pi) for this to happen automatically. The hardcoded partition numbers there match the hardcoded ones in the DT fstab overlay BTW.

At least it seems that you're now willing to accept the fact that it reads something from the device tree. So, instead of reading
```
/proc/device-tree/firmware/android/bootdevice
```
read
```
/proc/device-tree/firmware/android/fstab/boot/dev
/proc/device-tree/firmware/android/fstab/system/dev
/proc/device-tree/firmware/android/fstab/vendor/dev
/proc/device-tree/firmware/android/fstab/userdata/dev
```
and create the symlink if the partition matches.

This approach is something that I could possibly consider accepting but it also requires modifications elsewhere (i.e. introducing HACK to the DT fstab overlay).

**Comment by EatPrilosec** [comment]
>If you were to add the userdata partition to the DT fstab with the correct mnt and fs_mgr flags, Android doesn't even boot

i have to correct you again, i literally showed you a video of this not being true. specifying data in the DT fstab works fine, it overrides android's (not recovery's) fstab. i fnecessary i can make another video and show you the data partition gets mounted by first stage init (its verbose about it in the console while it boots up. if you pause the video i already provided, you can see the console talk about mounting data on early init.

**my current partition scheme has no mmcblk0p4**, becuase i have an extended partition, and the label is not userdata, so none of them would have booted, **android would literally fail to boot if DT-supplied userdata fstab did not work.** 


>This approach is something that I could possibly consider accepting but it also requires modifications elsewhere (i.e. introducing HACK to the DT fstab overlay).

thank you. i do appreciate it. and i have a suggestion (see below)

>It's HACK because we're not getting the boot device from Android bootloader (since we don't have one on Raspberry Pi) for this to happen automatically. The hardcoded partition numbers there match the hardcoded ones in the DT fstab overlay BTW.

correct, i understand that. but again, i have to correct you, dt OVERLAYS are not HARDCODED. theyre precompiled.. but not hardcoded, its an overlay. i can specify two of them in config.txt, the last one gets accepted, and the android-sdcard.dtbo gets ignored. theres nothing hardcoded about it on the Device Tree side, its just pre-supplied by the android build, and ignored by the OTA backup/restore script.

heres what i propose at the moment (untested sitll havent got around to compiling twrp

```
void create_device_symlinks(const Uevent& uevent, std::vector<std::string>& links)
{
    static const std::string bootdevice_path = "/proc/device-tree/firmware/android/bootdevice";

    static const std::string bootpart_path = "/proc/firmware/device-tree/android/fstab/boot/dev";
    static const std::string syspart_path = "/proc/firmware/device-tree/android/fstab/system/dev";
    static const std::string venpart_path = "/proc/firmware/device-tree/android/fstab/vendor/dev";
    static const std::string datapart_path = "/proc/firmware/device-tree/android/fstab/userdata/dev";

    std::string bootdevice;
    if (!ReadFileToString(bootdevice_path, &bootdevice)) {
        LOG(ERROR) << "create_device_symlinks: failed to read boot device from " << bootdevice_path;
    }

    std::string bootpart;
    if (!ReadFileToString(bootdevice_path, &bootpart)) {
       static const std::string bootpart = "NotDefined"; // i dont know if this is the corret way to define a var that has already been initiated in c++
    }
    std::string syspart;
    if (!ReadFileToString(syspart_path, &syspart)) {
        static const std::string syspart = "NotDefined";
    }
    std::string venpart;
    if (!ReadFileToString(venpart_path, &venpart)) {
        static const std::string venpart = "NotDefined";
    }
    std::string datapart;
    if (!ReadFileToString(datapart_path, &datapart)) {
        static const std::string datapart = "NotDefined";
    }


    LOG(INFO) << "create_device_symlinks: device " << uevent.device_name;
    if ((uevent.device_name.compare(bootpart) == 0)                                                       // when uevent device matches DT partition, make link? i think thats what i did lol
            || (uevent.device_name.compare("mmcblk0p1") == 0 && bootdevice.compare(0, 6, "sdcard") == 0) // otherwise keep current behaviour 
            || (uevent.device_name.compare("sda1") == 0 && bootdevice.compare(0, 3, "usb") == 0)
            || (uevent.device_name.compare("nvme0n1p1") == 0 && bootdevice.compare(0, 4, "nvme") == 0)) {
        links.emplace_back("/dev/block/by-name/boot");
    } else if ((uevent.device_name.compare(syspart) == 0) 
            || (uevent.device_name.compare("mmcblk0p2") == 0 && bootdevice.compare(0, 6, "sdcard") == 0)
            || (uevent.device_name.compare("sda2") == 0 && bootdevice.compare(0, 3, "usb") == 0)
            || (uevent.device_name.compare("nvme0n1p2") == 0 && bootdevice.compare(0, 4, "nvme") == 0)) {
        links.emplace_back("/dev/block/by-name/system");
    } else if ((uevent.device_name.compare(venpart) == 0)
            || (uevent.device_name.compare("mmcblk0p3") == 0 && bootdevice.compare(0, 6, "sdcard") == 0)
            || (uevent.device_name.compare("sda3") == 0 && bootdevice.compare(0, 3, "usb") == 0)
            || (uevent.device_name.compare("nvme0n1p3") == 0 && bootdevice.compare(0, 4, "nvme") == 0)) {
        links.emplace_back("/dev/block/by-name/vendor");
    } else if ((uevent.device_name.compare(datapart) == 0)
            || (uevent.device_name.compare("mmcblk0p4") == 0 && bootdevice.compare(0, 6, "sdcard") == 0)
            || (uevent.device_name.compare("sda4") == 0 && bootdevice.compare(0, 3, "usb") == 0)
            || (uevent.device_name.compare("nvme0n1p4") == 0 && bootdevice.compare(0, 4, "nvme") == 0)) {
        links.emplace_back("/dev/block/by-name/userdata");
    }
}

```

**Comment by EatPrilosec** [comment]
> At least it seems that you're now willing to accept the fact that it reads something from the device tree. So, instead of reading
> 
> ```
> /proc/device-tree/firmware/android/bootdevice
> ```
> 
> read
> 
> ```
> /proc/device-tree/firmware/android/fstab/boot/dev
> /proc/device-tree/firmware/android/fstab/system/dev
> /proc/device-tree/firmware/android/fstab/vendor/dev
> /proc/device-tree/firmware/android/fstab/userdata/dev
> ```
> 
> and create the symlink if the partition matches.

just to clarify:

1)
> /proc/device-tree/firmware/android/fstab/**_userdata_**/dev
should be `/proc/device-tree/firmware/android/fstab/data/dev` [based off the dts i provided](https://github.com/twrp-raspberry/android_local_manifest/issues/2#issuecomment-2380053064)

2)

userdata is not ALWAYS defined, so it would likely be best practice to keep current behavior as a backup.

**Comment by EatPrilosec** [comment]
> LMFAO, this is now down to semantics of a term "hardcoded"?

you being wrong isnt a semantic argument. 

have fun being super fun at parties bro.

**Comment by KonstaT** [comment]
> have fun being super fun ...

What's the semantics of this?

> thats cool, i ran out of crayons trying to explain to you the rpi boot process youre clearly unfamiliar with.

I run out of crayons trying to explain to you the Android boot process you're clearly unfamiliar with.

In any case, good luck with your future endeavours and everything else in life. :)

**Comment by EatPrilosec** [comment]
> > have fun being super fun ...
> 
> What's the semantics of this?

there is nothin grammatically or semantically incorrect about that insult directed towards you, cry harder.

> > thats cool, i ran out of crayons trying to explain to you the rpi boot process youre clearly unfamiliar with.
> 
> I run out of crayons trying to explain to you the Android boot process you're clearly unfamiliar with.

if only more than one of use had humility. I've already expressed my unfamiliarity. unlike someone here. 


> In any case, good luck with your future endeavours and everything else in life. :)

endeavors*

cheers mate.

**Comment by KonstaT** [comment]
> endeavors*

https://dictionary.cambridge.org/dictionary/english/endeavour

**Comment by EatPrilosec** [comment]
> I don't have time/energy/interest to partake any further.

unless your _feelie weelies_ get hurt i suppose.....


> https://dictionary.cambridge.org/dictionary/english/endeavour

https://www.grammar.com/endeavor_vs._endeavour

typical brit. github issues are not formal.

cope.

**Comment by KonstaT** [comment]
> https://www.grammar.com/endeavor_vs._endeavour

LOL, then why "correct" it if it wasn't misspelled in the first place.

Since I don't need to provide you a platform to continue your insult, enjoy being blocked.
